crates/services/upgradable-executor/src/executor.rs

```diff
@@ -339,46 +345,89 @@ where
     }
 
     #[cfg(feature = "wasm-executor")]
-    fn execute_inner<TxSource>(
+    fn produce_inner<TxSource>(
         &self,
-        block: ExecutionBlockWithSource<TxSource>,
+        block: Components<TxSource>,
         options: ExecutionOptions,
     ) -> ExecutorResult<Uncommitted<ExecutionResult, Changes>>
     where
         TxSource: TransactionsSource + Send + Sync + 'static,
     {
-        let block_version = block.state_transition_version();
+        let block_version = block.header_to_produce.state_transition_bytecode_version;
         let native_executor_version = self.native_executor_version();
         if block_version == native_executor_version {
             match &self.execution_strategy {
-                ExecutionStrategy::Native => self.native_execute_inner(block, options),
+                ExecutionStrategy::Native => self.native_produce_inner(block, options),
                 ExecutionStrategy::Wasm { module } => {
-                    self.wasm_execute_inner(module, block, options)
+                    self.wasm_produce_inner(module, block, options)
                 }
             }
         } else {
             let module = self.get_module(block_version)?;
-            tracing::warn!(
-                "The block version({}) is different from the native executor version({}). \
-                The WASM executor will be used.", block_version, Self::VERSION
-            );
-            self.wasm_execute_inner(&module, block, options)
+            Self::trace_block_version_warning(block_version);
+            self.wasm_produce_inner(&module, block, options)
         }
     }
 
     #[cfg(not(feature = "wasm-executor"))]
-    fn execute_inner<TxSource>(
+    fn produce_inner<TxSource>(
         &self,
-        block: ExecutionBlockWithSource<TxSource>,
+        block: Components<TxSource>,
         options: ExecutionOptions,
     ) -> ExecutorResult<Uncommitted<ExecutionResult, Changes>>
     where
         TxSource: TransactionsSource + Send + Sync + 'static,
     {
-        let block_version = block.state_transition_version();
+        let block_version = block.header_to_produce.state_transition_bytecode_version;
         let native_executor_version = self.native_executor_version();
         if block_version == native_executor_version {
-            self.native_execute_inner(block, options)
+            self.native_produce_inner(block, options)
+        } else {
+            Err(ExecutorError::Other(format!(
+                "Not supported version `{block_version}`. Expected version is `{}`",
+                Self::VERSION
+            )))
+        }
+    }
+
+    #[cfg(feature = "wasm-executor")]
+    fn dry_run_inner<TxSource>(
```

Comment:
Okay. I just added a `dry_run` boolean param to some of the internal functions to branch. I don't think we need a type. I kept the external functions the same, partly because we have a bunch of test functions that won't ever be used for dry_run, so I didn't want to clutter the interfaces with the extra param.